### PR TITLE
feat(data-warehouse): add the ability to connect to MySQL as a datawarehouse source

### PR DESCRIPTION
**Pull request by jspizziri**
## Problem

Currently, the data warehouse functionality supports Postgres. This PR seeks to expand that functionality to allow for the import of MySQL/MariaDB databases.

## Changes

posthog-dw-mysql
posthog-dw-mysql-add

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

I tested the code by adding a MySQL flavored DB via the UI. However, it seems that there are some dependencies on a connection to an S3 bucket that I don't have. I'm largely looking for two things:

1. I'd like to see if the Posthog team would consider adding this functionality/merging this PR.
2. If the answer to the previous question is "yes", I'd like to ask for high-level guidance on this PR, as I have some questions.

### CODE REVIEW

**Comment by jspizziri** [comment]
Still waiting on input from a contributor.

**Comment by jspizziri** [comment]
@EDsCODE, sorry for pinging you here, but you seem to be the contributor that's most active in the space of the datawarehouse. I'd love some feedback from a contributor on this PR.

**Comment by jspizziri** [comment]
@EDsCODE @benjackwhite @pauldambra could anyone chime in here? I'm assuming that this is getting lost in the shuffle due to perhaps this project not having regular external contributors?

File: posthog/warehouse/models/external_data_schema.py [hunk 1]
```diff
@@ -141,3 +154,64 @@ def get_schemas(postgres_host: str, postgres_port: int):
...
+        connection = pymysql.connect(
+            host=mysql_host,
+            port=mysql_port,
+            database=database,
+            user=user,
+            password=password,
+            connect_timeout=5,
+            # ssl_ca="/tmp/no.txt",
+            # ssl_cert="/tmp/no.txt",
+            # ssl_key="/tmp/no.txt",
```
Comment (Gilbert09) [inline posthog/warehouse/models/external_data_schema.py:177]: Nit: these can be removed

**Review by Gilbert09 — COMMENTED** [review]
This looks great, thank you for adding this. If you could get `master` merged in to this branch, and I'll do some testing locally 🚀

**Comment by jspizziri** [comment]
@Gilbert09 thanks for your response and review! I'm working on getting things rebased and re-tested right now. I'll let you know when its ready for another review.

**Comment by jspizziri** [comment]
@Gilbert09 , alright. I've rebased on `master` and I've gotten things back to where I had them previously (which was a failure due to not having access to an S3 bucket). The rebase was a bit of a mess as a good deal of stuff has changed since my initial implementation. I _think_ things are correct-ish at this point.

I implemented your feedback in an isolated commit for your review https://github.com/PostHog/posthog/pull/23370/commits/2692d2a9203131c342bd9dba23a4280a2a1d8518 and I can squash it into the main commit once you've had a chance to look at it.

Hopefully things work for you!

**Comment by jspizziri** [comment]
@Gilbert09 sorry for bugging you. Work on the DW stuff is happening so fast that my PR keeps getting new merge conflicts every few days. It'd be great to get this tested and merged so that I don't need to keep rebasing every few days to keep things up-to-date. I'm going to rebase to resolve the most recent batch of conflicts now. Thanks again for looking at it!

**Comment by Gilbert09** [comment]
> @Gilbert09 sorry for bugging you. Work on the DW stuff is happening so fast that my PR keeps getting new merge conflicts every few days. It'd be great to get this tested and merged so that I don't need to keep rebasing every few days to keep things up-to-date. I'm going to rebase to resolve the most recent batch of conflicts now. Thanks again for looking at it!

@jspizziri Sorry for the delay - I'm looking at this now

**Comment by Gilbert09** [comment]
@jspizziri I needed to make a few minor changes and didn't have permission to push to your repo. So I've opened a new PR with my changes on top here - https://github.com/PostHog/posthog/pull/23921 - I'll close this in favor of that and get it merged once CI has passed ✅ Thank you for your contribution!

**Comment by jspizziri** [comment]
@Gilbert09 , you're a hero! Thanks for your help on this. P.S. I was intending to squash my "[implement review feedback](https://github.com/PostHog/posthog/pull/23921/commits/402c7cdf97ef670d76cbfd1f8456e8c3e6874bc9)" commit into my main one, but now that you've forked it I won't be able to do that. LMK if you need me to do any updates.

**Comment by Gilbert09** [comment]
@jspizziri This is now live on PostHog! Thank you again for your contribution

**Pull request closed without merging**